### ISSUE TITLE
`ConnectionTransactionContextManager` slowly drains the pool

### ISSUE THREAD

**Issue by villlem** [CanopyTax/asyncpgsa#52]
We are using transaction as follows

```python
async with pool.transaction() as conn:
    ...
```

and from time to time (in our use case it ranges from once a day to once a week) our service freezes. When we investigated the issue we noticed, that all connections in the pool were marked as used and no free connection was available. We traced the problem to the `ConnectionTransactionContextManager`. 

If cancellation or timeout raises during `__aenter__`  or `__aexit__` there is no guarantee that connection is returned to the pool. And it slowly drains. We confirmed that this is the issue, because we change code to this

```python
async with pool.acquire() as conn:
    async with conn.transaction():
        ....
```

and our problem stops.

I can create PR to fix this if you like.

**Comment by nhumrich** [comment]
Wow awesome find. A PR would be greatly helpful

**Comment by bitrut** [comment]
I've been using asyncpg singleton and ran into the same issue under load tests of my service. Pool size is 10 and after some time all the connections in the pool are become useless.

@villlem Did you manage to fix it?

**Comment by villlem** [comment]
Not yet. Unfortunately my time is short, but I plan to do it next month.

Dne út 9. 1. 2018 16:12 uživatel Paweł Kowalski <notifications@github.com>
napsal:

**Comment by nhumrich** [comment]
@villlem any update? It would be great to get this fixed. If you dont have time to pr, do you have a way to reproduce it so that I can dig further?

**Comment by villlem** [comment]
I'm so sorry I know how fix this but I dont' have time do this (properly). I dont have means to test it. Nobody allow me to deploy potentialy broken code to production (because) only there this problem occurs.
What I recomend to do is by default shield (https://docs.python.org/3/library/asyncio-task.html#asyncio.shield) entire `__aenter__` and `__aexit__` methods so they are not interupted by cancellation.

**Comment by nhumrich** [comment]
cool. Thanks for the recommendations.

**Comment by nhumrich** [comment]
Hopefully this is fixed in 0.19.2. I dont really have a good way to test this/reproduce it. So would be good if others can verify if its broken still or not.
